apis/v1alpha2/gateway_types.go

```diff
@@ -0,0 +1,835 @@
+
+	// This reason is used with the "Ready" condition when one or
+	// more Listeners are not ready to serve traffic.
+	GatewayReasonListenersNotReady GatewayConditionReason = "ListenersNotReady"
+
+	// This reason is used with the "Ready" condition when the requested
+	// address has not been assigned to the Gateway. This reason
+	// can be used to express a range of circumstances, including
+	// (but not limited to) IPAM address exhaustion, invalid
+	// or unsupported address requests, or a named address not
+	// being found.
+	GatewayReasonAddressNotAssigned GatewayConditionReason = "AddressNotAssigned"
+)
+
+// ListenerStatus is the status associated with a Listener.
+type ListenerStatus struct {
+	// Name is the name of the Listener. If the Gateway has more than one
+	// Listener present, each ListenerStatus MUST specify a name. The names of
+	// ListenerStatus objects MUST be unique within a Gateway.
+	//
+	// +kubebuilder:validation:MaxLength=253
+	Name string `json:"name"`
+
+	// SupportedKinds is the list indicating the Kinds supported by this
+	// listener. When this is not specified on the Listener, this MUST represent
+	// the kinds an implementation supports for the specified protocol. When
+	// there are kinds specified on the Listener, this MUST represent the
+	// intersection of those kinds and the kinds supported by the implementation
+	// for the specified protocol.
+	//
+	// +kubebuilder:validation:MaxItems=8
+	SupportedKinds []RouteGroupKind `json:"supportedKinds"`
+
+	// AttachedRoutes represents the total number of Routes that have been
+	// successfully attached to this Listener.
+	AttachedRoutes int32 `json:"attachedRoutes"`
+
+	// Conditions describe the current condition of this listener.
+	//
+	// +listType=map
+	// +listMapKey=type
+	// +kubebuilder:validation:MaxItems=8
+	Conditions []metav1.Condition `json:"conditions"`
+}
+
+// ListenerConditionType is a type of condition associated with the
+// listener. This type should be used with the ListenerStatus.Conditions
+// field.
+type ListenerConditionType string
+
+// ListenerConditionReason defines the set of reasons that explain
+// why a particular Listener condition type has been raised.
+type ListenerConditionReason string
+
+const (
+	// This condition indicates that the controller was unable to resolve
+	// conflicting specification requirements for this Listener. If a
+	// Listener is conflicted, its network port should not be configured
+	// on any network elements.
+	//
+	// Possible reasons for this condition to be true are:
+	//
+	// * "HostnameConflict"
+	// * "ProtocolConflict"
+	// * "RouteConflict"
+	//
+	// Possible reasons for this condition to be false are:
+	//
+	// * "NoConflicts"
+	//
+	// Controllers may raise this condition with other reasons,
+	// but should prefer to use the reasons listed above to improve
+	// interoperability.
+	ListenerConditionConflicted ListenerConditionType = "Conflicted"
```

Comment:
We actually had quite a bit of conversation around polarity of conditions here. As I recall, the we ended up settling on the condition names that felt most natural, since the alternative of consistent polarity always resulted in at least one confusing name. @jpeach Probably has more context on this. I think both `Consistent` and `Attached` were the alternatives we explored for those conditions, but we didn't like the potential confusion around the `Consistent` name.

Reply:
`Ready` could not be true on a Listener when `Conflicted` or `Detached` are true. `ResolvedRefs` is more of a grey area; an implementation could conceivably make a listener `Ready` without resolving all the references (e.g. by serving error responses).

In the discussion around naming the conditions, we ended up using "natural polarity". The API conventions used to recommend negative polarity, but now don't take a position.